### PR TITLE
Rename @alpha to @targetName

### LINKED ISSUE #7942: @alpha has no effect over double definition error
## minimized code

```Scala
import scala.annotation.alpha
@alpha("fooString") def foo(ps: String*) : Unit = ???
@alpha("fooInt") def foo(ps: Int*) : Unit = ???
@alpha("fooAny") def foo(ps: Any*) : Unit = ???
```

<details>
<summary>Compilation output</summary>

```scala
Double definition:
def foo(ps: String*): Unit in package object main$package at line 4 and
def foo(ps: Int*): Unit in package object main$package at line 5
have the same type after erasure.
@alpha("fooAny") def foo(ps: Any*) : Unit = ???

Double definition:
def foo(ps: Int*): Unit in package object main$package at line 5 and
def foo(ps: Any*): Unit in package object main$package at line 6
have the same type after erasure.
```
</details>

see scastie: https://scastie.scala-lang.org/owQEIEShQayiRjUJ6Esk2w

## expectation
Currently `DummyImplicit` is used as a workaround to create unique signatures after erasure. 
@odersky [mentioned](https://contributors.scala-lang.org/t/principles-for-implicits-in-scala-3/3072/126?u=soronpo) that `@alpha` annotation could be used to avoid double definition error, however it does not seem to affect the compiler when it checks for double definitions.

### PR DESCRIPTION
**Pull request by odersky**
... and make @targetName work to avoid double definitions. Two definitions now `match` only if they have the same names, signatures and erased names. The erased name of a definition is the name given in a `@targetName` annotation, or else
the declared name, if no `@targetName` annotation is given.

Fixes #7942

### CODE REVIEW

**Comment by soronpo** [comment]
Does this change have any effect over the possibility solving https://github.com/lampepfl/dotty/issues/7942 ?

**Comment by odersky** [comment]
@soronpo Yes, this fixes #7942.

File: docs/docs/reference/other-new-features/targetName.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+Here, the original methods `g` and `f` do not override each other since they have
+different names. But once we switch to target names, there is a clash that is reported by the compiler:
+```
+-- [E120] Naming Error: test.scala:4:6 -----------------------------------------
+4 |class B extends A:
+  |      ^
+  |      Name clash between defined and inherited member:
+  |      def f(): Int in class A at line 3 and
+  |      def g(): Int in class B at line 5
+  |      have the same name and type after erasure.
```
Comment (soronpo) [inline docs/docs/reference/other-new-features/targetName.md:77]: Regarding the error message, maybe it's worth proposing the user the option of using `@targetName` to resolve the error.

**Comment by soronpo** [comment]
BTW, is there a purpose for `DummyImplicit` now, or can it be deprecated? AFAIK, it was only used to define unique signatures.

**Comment by odersky** [comment]
I think `DummyImplicit` could be deprecated. But it lives in 2.13, so for now it's untouchable :-)

**Review by liufengyun — APPROVED** [review]
LGTM, with some minor comments.

The following code now compiles:

```Scala
import scala.annotation.targetName

class B {
  final def foo(): Int = ???
  def bar(): Int = ???
}

class C extends B {
  @targetName("fooB") def foo(): Int = ???
  @targetName("barB") def bar(): Int = ???
}
```

Instead, I think the compiler should report two errors above. The reason is that `@targetName` is intended to overcome the limitations of target platforms, rather than change the (static) semantics of the surface language. 

The following code currently compiles, it should be rejected:

```Scala
class C {
  @targetName("foo B") def foo(): Int = ???
}
```

The following code currently compiles, should be rejected:

```Scala
class C(x: String) {
  @targetName(x) def foo(): Int = ???
}
```


/cc @bishabosha Could you please comment about Scala 2 forward interoperability?

The following comment is intended for open discussions in the future, not for this PR.

## explicit annotation VS synthesized annotation

The advantage of explicit annotation is obvious: better debuggability and Java interoperability.

On the other hand, when programmers don't care about Java interoperability & debuggability (e.g. during prototyping), can the compiler automatically synthesize target names to overcome the limitations of the target platform? 

There are already many places where the compiler performs synthesis to overcome platform limitations, is it possible to hide the implementation detail for `targetName` by default, and allows users to optionally use `@targetName` when they care about debuggability and Java Interop?

**Comment by sjrd** [comment]
> The following code currently compiles, it should be rejected:
> 
> ```scala
> class C {
>   @targetName("foo B") def foo(): Int = ???
> }
> ```

I don't think so. I believe the only characters that should be rejected are `. ; / [ <`, as those are the characters that are not valid in a JVM method name. For classes we should also allow `<`.

**Comment by soronpo** [comment]
> 
> 
> > The following code currently compiles, it should be rejected:
> > ```scala
> > class C {
> >   @targetName("foo B") def foo(): Int = ???
> > }
> > ```
> 
> I don't think so. I believe the only characters that should be rejected are `. ; / [ <`, as those are the characters that are not valid in a JVM method name. For classes we should also allow `<`.
> 
> We should forbid unpaired surrogate characters as well.

But Scala 2 allows ```def `.;/[<` : Unit = {} ```. What happens at the backend in this case?

**Comment by smarter** [comment]
These characters get encoded: https://github.com/scala/scala/blob/0a1942c82f18f11cf6d39d0a1e300e9995dcdbfc/src/library/scala/reflect/NameTransformer.scala#L45-L62

**Comment by liufengyun** [comment]
> > The following code currently compiles, it should be rejected:
> > ```scala
> > class C {
> >   @targetName("foo B") def foo(): Int = ???
> > }
> > ```
> 
> I don't think so. I believe the only characters that should be rejected are `. ; / [ <`, as those are the characters that are not valid in a JVM method name. For classes we should also allow `<`.
> 
> We should forbid unpaired surrogate characters as well.

Good point, as Scala already allows such names with backquotes, for consistency they need to be supported (However, for better platform portability, it seems better for language design decisions to be independent of a particular platform).

**Comment by sjrd** [comment]
> (However, for better platform portability, it seems better for language design decisions to be independent of a particular platform).

In practice, the other platforms already have to handle what the JVM can handle.

**Comment by odersky** [comment]
> I don't think so. I believe the only characters that should be rejected are . ; / [ <, as those are the characters that are not valid in a JVM method name. For classes we should also allow <.

> We should forbid unpaired surrogate characters as well.

The backend encodes names so that illegal method names are avoided. For instance,
```scala
def `..` = 1
```
will end up as 
```
public int $u002E$u002E();
```
`@targetName` should behave exactly like backticks. Either they both admit all idents and avoid bad ones by encodings, or none of them should do it. So this is a separate issue.

**Comment by odersky** [comment]
@liufengyun 

> Instead, I think the compiler should report two errors above. The reason is that @targetName is intended to overcome the limitations of target platforms, rather than change the (static) semantics of the surface language.

That's quite a fundamental question we need to answer here. Should @targetName be allowed to refine the overriding behavior? Right now the answer is yes. If we change that it could get tricky and also expensive. Here's a problematic example:
```scala
trait A:
  def f: Any
trait B:
  @targetName("ff") def f: Int
abstract class C extends A, B
```
That's currently legal. If it should be illegal, this means that that no class could ever extend both `A` and `B`. Even if we tried to add implementations to class `C` like this:
```
class C extends A, B:
  override def f: Any = ""
  @targetName("ff") override def f: Int = 1
```
it still would not work since each of the two definitions of `f` in `C` "overrides" one other instance with a different target name.

So this would be a pretty harsh restriction, and it's particular annoying since it arises from global behavior and cannot be controlled by local knowledge alone. 

The other objection is that even checking a situation like this is expensive. Enumerating overriding pairs is quadratic in the size of the program. It's a significant part of total compile-time when compiling things like stdlib. Here we would make it twice as expensive. 

On the other hand, what would we really gain from a restriction?

**Comment by odersky** [comment]
> The following code currently compiles, should be rejected:
```scala
class C(x: String) {
  @targetName(x) def foo(): Int = ???
}
```
Is there a way to do this by annotating the definition of the `targetName` class in some way? I.e. can we force annotations to take only String literals as parameters?

**Comment by liufengyun** [comment]
> @liufengyun
> 
> > Instead, I think the compiler should report two errors above. The reason is that @targetName is intended to overcome the limitations of target platforms, rather than change the (static) semantics of the surface language.
> 
> That's quite a fundamental question we need to answer here. Should @targetName be allowed to refine the overriding behavior? Right now the answer is yes. If we change that it could get tricky and also expensive. Here's a problematic example:
> 
> ```scala
> trait A:
>   def f: Any
> trait B:
>   @targetName("ff") def f: Int
> abstract class C extends A, B
> ```
> 
> That's currently legal. If it should be illegal, this means that that no class could ever extend both `A` and `B`. Even if we tried to add implementations to class `C` like this:
> 
> ```
> class C extends A, B:
>   override def f: Any = ""
>   @targetName("ff") override def f: Int = 1
> ```
> 
> it still would not work since each of the two definitions of `f` in `C` "overrides" one other instance with a different target name.

To clarify I think the following example will work:

```Scala
trait A {
  def f: Any
}

trait B {
  @targetName("ff")
  def f: Int
}

abstract class C extends A with B {
  def f: Any = ""
  @targetName("ff")
  def f: Int = 3
}
```

The reason is that _overriding check_ takes the **whole type** into consideration, while _conflicit check_ only takes **signature** and **target name** into consideration.


> 
> So this would be a pretty harsh restriction, and it's particular annoying since it arises from global behavior and cannot be controlled by local knowledge alone.
> 
> The other objection is that even checking a situation like this is expensive. Enumerating overriding pairs is quadratic in the size of the program. It's a significant part of total compile-time when compiling things like stdlib. Here we would make it twice as expensive.

What I propose here is the following:

- In overriding check, use original names and in addition make sure that overridding pairs have the same target names
- In conflict member check, use target names.

With the scheme above, will performance be less a concern?

> 
> On the other hand, what would we really gain from a restriction?

I think it is safer to guard against abuse of language features:

```Scala
class A {
  final def f(): Int = ???
}

class B extends A {
  @targetName("ff")
  def f(): Int = ???
}
```

**Comment by odersky** [comment]
> The reason is that overriding check takes the whole type into consideration, while conflicit check only takes signature and target name into consideration.

No that does not work. `()Int` is a legal override of `()Any`.

**Comment by liufengyun** [comment]
> > The reason is that overriding check takes the whole type into consideration, while conflicit check only takes signature and target name into consideration.
> 
> No that does not work. `()Int` is a legal override of `()Any`.

Yes, I missed that `Int <: Any`. The essential question seems to be

1. Only use the feature to remove platform restrictions, or
2. Also use it to make the surface language more expressive

The use cases intended by the first is clear, while it is unclear for the 2nd. Another worry about the 2nd is that it may create inconsistencies. For example, can we override the target name directly?

```Scala
trait A {
  @targetName("ff")
  def f(): Int
}

class B extends A {
  def ff(): Int = ???
}
```

In the current PR, the code above is rejected, which makes reasoning about overriding more complex.

**Comment by odersky** [comment]
> In the current PR, the code above is rejected, which makes reasoning about overriding more complex.

Here's what's stated in targetName.md:

`@targetName` annotations are significant for matching two method definitions to decide whether they conflict or override each other. Two method definitions match if they have the same name, signature, and erased name. Here,

 - The _signature_ of a definition consists of the names of the erased types of all (value-) parameters and the method's result type.
 - The _erased name_ of a method definition is its target name if a `@targetName`
   annotation is given and its defined name otherwise.

So, both regular names and target names need to match.

**Comment by bishabosha** [comment]
The changes to Tasty should be fine in this case for Scala 2 forward compat, it makes it easier than special casing the `alpha` annotation

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
@liufengyun We now avoid the power of @targetName where it deviates from normal semantics. The new rules are as follows:

 - Two members can override each other if their names and signatures are the same,
   and they either have the same erased names or the same types.
 - If two members override, then both their erased names and their types must be the same. their own. 

So the means that targetNames can only avoid breakage due to erasure clashes. They cannot introduce new distinctions on their own. 

I also found a way to make this efficient. The `OverridingPairs` cursor in `RefChecks` implements exactly the behavior des ribed above. Then we check for each returned pair in `checkOverride` that both erased names and types match. So we still need only a single pass to do this.

@liufengyun Can you review the latest commits implementing this?

**Review by liufengyun — APPROVED** [review]
LGTM

**Pull request merged**